02-microarray/pathway-analysis_microarray_04_gsva.Rmd

```diff
@@ -0,0 +1,631 @@
+
+The function we will use to map from Ensembl gene IDs to Entrez gene IDs is called `mapIds()`.
+
+Let's create a data frame that shows the mapped Entrez IDs along with the gene expression values for the respective Ensembl IDs.
+
+```{r}
+# First let's create a mapped data frame we can join to the gene expression values
+mapped_df <- data.frame(
+  "entrez_id" = mapIds(
+    org.Hs.eg.db, # Replace with annotation package for the organism relevant to your data
+    keys = df$Gene,
+    column = "ENTREZID", # Replace with the type of gene identifiers you would like to map to
+    keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
+    multiVals = "first" # This will keep only the first mapped value for each Ensembl ID
+  )
+) %>%
+  # If an Ensembl gene identifier doesn't map to a Entrez gene identifier, drop that
+  # from the data frame
+  dplyr::filter(!is.na(entrez_id)) %>%
+  # Make an `Ensembl` column to store the rownames
+  tibble::rownames_to_column("Ensembl") %>%
+  # Now let's join the rest of the expression data
+  dplyr::inner_join(df, by = c("Ensembl" = "Gene"))
+```
+
+This `1:many mapping between keys and columns` message means that some Ensembl gene identifiers map to multiple Entrez IDs.
+In this case, it's also possible that a Entrez ID will map to multiple Ensembl IDs.
+For the purpose of performing GSEA later in this notebook, we keep only the first mapped IDs.
+For more about how to explore this, take a look at our [microarray gene ID conversion example](https://alexslemonade.github.io/refinebio-examples/02-microarray/gene-id-annotation_microarray_01_ensembl.html).
+
+Let's see a preview of `mapped_df`.
+
+```{r}
+head(mapped_df)
+```
+
+We will want to keep in mind that GSVA requires that data is in a matrix with the gene identifiers as rownames.
+In order to successfully turn our data frame into a matrix, we will need to ensure that we do not have any duplicate gene identifiers.
+
+Let's check to see if we have any Entrez IDs that mapped to multiple Ensembl IDs.
+
+```{r}
+any(duplicated(mapped_df$entrez_id))
+```
+
+Looks like we do have duplicated Entrez IDs.
+Let's find out which Entrez IDs have been duplicated.
+
+```{r}
+dup_entrez_ids <- mapped_df %>%
+  dplyr::filter(duplicated(entrez_id)) %>%
+  dplyr::pull(entrez_id)
+
+dup_entrez_ids
+```
+
+Now let's take a look at the rows associated with the duplicated Entrez IDs.
+
+```{r}
+mapped_df %>%
+  dplyr::filter(entrez_id %in% dup_entrez_ids)
+```
+
+We can see that the associated values vary for each row.
+
+As we mentioned earlier, we will want to remove duplicated gene identifiers in preparation for the GSVA steps later, so let's keep the Entrez IDs associated with the larger mean expression value per sample.
```

Comment:
This seems like a somewhat standard workflow for RNA-seq data because you might expect lowly expressed genes not to show up in your data, so you'd drop the instance of the identifier with low expression as estimated by the mean value across samples. For GSEA (not the pre-ranked version), the default way to collapse identifiers is to take the max value for multiple probes that map to  but you can apply the same idea to Ensembl IDs that map to the same Entrez ID ([ref](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html)). Other ways people do it (also at that ref) are to take the mean value or median value (should be less sensitive to outliers than the mean). Either way, it makes more sense to me to work with the _individual values_ (e.g., not based on a summary measurement across all samples) given that you're doing GSVA which is on a per sample basis. If you pick the max value, you're preferentially going to pick something with a higher rank and that caveat should get documented.

Reply:
Gotcha, that all makes sense. I will use the default way and take the max value and be sure to document the caveat (as well as link to the [reference](https://www.gsea-msigdb.org/gsea/doc/GSEAUserGuideFrame.html)).